components/server/src/prometheus-metrics.ts

```diff
@@ -55,6 +55,18 @@ export function increaseApiCallCounter(method: string, statusCode: number) {
     apiCallCounter.inc({ method, statusCode });
 }
 
+export const apiCallDurationHistogram = new prometheusClient.Histogram({
+    name: 'gitpod_server_api_calls_duration_seconds',
+    help: 'Duration of API calls in seconds',
+    labelNames: ['method'],
+    buckets: [0.01, 0.05, 0.1, 0.5, 1, 5, 10],
```

Comment:
This looks fine for starters. Let's have a look at the current numbers in prod. I can imagine that especially in 0.5..5 we want more resolution; but that's pure speculation for now.

Reply:
Yes I thought so too. But for showing the quantiles in Grafana, I'm getting NaNs for this query `histogram_quantile(0.9, sum (rate(gitpod_server_api_calls_duration_seconds_bucket[30m])) by (method, le))`. You can see histogram_quantile(0.9, sum (rate(gitpod_server_api_calls_duration_seconds_bucket[30m])) by (method, le)).

Reply:
Hm. I'm still thinking about how to best display this data. :thinking: Will have a look again tomorrow.

Reply:
Yes no worries, still a rough WIP, will continue to play around. Thanks for looking 🤗

Reply:
@laushinka I took another look: the core "problem" here is our "method" dimension and it's high cardinality. Because of that I think it makes sense to have two things:
 - a heatmap: this works very well to get a sense of development over time _for a single method_. With multiple methods this gets unreadable instantly
 - a percentile view ([openvsx proxy example](https://grafana.gitpod.io/d/HNOvmGpxgd/openvsx-proxy?orgId=1&editPanel=8)): this is not as precise, but might be a little bit better when comparing multiple methods/overall situation

Sadly I could not test atm because prometheus stopped returning data :confused: I hope it works for you; please ping on Slack if you face problems during development here. This should be way smoother I think.

Reply:
Thanks, I was just testing with heatmap[[1](https://grafana-laushinka-histogram-7117.preview.gitpod-dev.com/d/server/gitpod-component-server?orgId=1&viewPanel=66&from=1641982059914&to=1641985659914)]! And you're right, when I use heatmap, the method label becomes unusable. Will check out the percentile view as well. I'll ping on Slack.